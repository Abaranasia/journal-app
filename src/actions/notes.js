import { db } from "../firebase/firebaseConfig";
import { types } from "../types/types";

import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";


export const startNewNote = () => { // Acción para crear una nueva nota
    return async (dispatch, getState) => { //comunicación asíncrona con Thunk 

        const uid = getState().auth.uid;
        //console.log(uid)

        const newNote = {
            title: 'Otra nueva nota',
            body: 'jujijaje',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote); // referencia al documento en Firestore donde añadir una nueva nota para este usuario
        // console.log(docRef)
        dispatch(activeNote(docRef.id, newNote))
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }

})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => { // acción asíncrona porque depende del backend

        const notes = await loadNotes(uid); // obtenemos todas las notas del usuario identificado y las enviaremos al store
        dispatch(setNotes(notes)) //hacemos el dispatch para enviar el listado de notas al store mediante la acción setNotes
    }
}


export const setNotes = (notes) => ({ // Acción para enviar al store las notas recibidas de Firebase
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => { // Acción para guardar una nota en la BBDD
    // necesitaremos el middleware Thunk para poder comunicarnos con Firebase
    return async (dispatch, getState) => { //Thunk, que es asíncrono, devuelve el dispatch y el uid del usuario

        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url; //Para evitar que Firebase pete si le llega un undefined
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id; // Hacemos esto para extraer la id de note, dado que no la necesitamos en la inserción

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore); // Podemos enviar noteToFirestore o note, pro la primera no lleva id, ojo

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved!', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({ // Acción que permite refrescar una nota, en caso de que haya cambiado, por ejemplo, tras guardarla
    type: types.notesUpdated,
    payload: {
        id,
        note: {  // Tenemos que pasarlo así porque este note nos llega sin id al enviar noteToFirestore, asi que lo volvemos a poner
            id, //Sin este este id, jornalEntry nos da un error porque debe dibujar un elemento sin key
            ...note
        }

    }
})

