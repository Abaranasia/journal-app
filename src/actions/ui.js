import { types } from '../types/types';

/** Este mÃ³dulo define los actions en referencia a los error del sistema */

export const setError = (err) => ({
    type: types.uiSetError,
    payload: err
})

export const removeError = () => ({
    type: types.uiRemoveError
})

export const startLoading = () => ({
    type: types.uiStartLoading,

})

export const finishLoading = () => ({
    type: types.uiFinishLoading,

})