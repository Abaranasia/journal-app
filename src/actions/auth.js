import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import { notesLogout } from "./notes";


/** Este módulo define los actions para el proceso de auth.
 *  Aquí incluiríamos todos los métodos de acceso necesarios: email, google, twitter, etc.
 */


export const startLoginEmailPassword = (email, password) => {
    // middleware para login con email que gestiona el action de login

    return (dispatch) => { //devuelve un objeto llamado dispatch con el tipo y el payload

        dispatch(startLoading()); // Empieza el procedimiento de login

        
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
            })
            .catch(e => {
                console.log('Hubo un error en el login:', e);
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
                Swal.fire('Error', e.message, 'error')
            });
/*         setTimeout(() => {
            dispatch(login(123, 'Pedro'))
        }, 3500); */

    }
}

export const startRegisterWithEmailPassword = (email, password, name) => {
    // dado que es ua tarea asíncrona, necesito retornar un callback
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name }); // hacemos esto porque el registro en Firebase no lo hace solo
                console.log(user);

                dispatch(
                    login(user.uid, user.displayName)
                )
            })
            .catch(e => { console.log('Hubo un error al registrarse:', e) });
    }
}

export const startGoogleLogin = () => {
    // middleware para login con usuario de Google que gestiona el action de login

    return (dispatch) => {
    //devuelve una función asíncrona que recibe un dispatch con el tipo y el payload que deben ser enviados

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                //console.log(user.uid, user.displayName);
                dispatch(
                    login(user.uid, user.displayName)
                )
            });
    }
}


export const login = (uid, displayName) => {
    //action para login: devuelve un tipo (login) y un payload con uid y nombre

    return {
        type: types.login,
        payload: {
            uid, displayName
        }
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout())
        dispatch(notesLogout())
    }
}
export const logout = (params) => ({
    type: types.logout
})
