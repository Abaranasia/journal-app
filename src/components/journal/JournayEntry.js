import React from 'react';
import moment from "moment";
import { activeNote } from '../../actions/notes';
import { useDispatch } from 'react-redux';

export const JournayEntry = ({ id, date, title, body, url }) => {

    const noteDate = moment(date);
    const dispatch = useDispatch();

    const handleEntryClick = () => {
        dispatch(activeNote(id, { //Env√≠a al state del store la nota seleccionada y la define como activa
            date, title, body, url
        }))
    }

    return (
        <div className="journal__entry pointer" onClick={handleEntryClick}>
            <div
                className="journal__entry-picture"
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: 'url(https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQVVkkZjXvk82ffyCBUJOEtt8fNbVpLUh2mpw&usqp=CAU)'
                    //backgroundImage: `url(${url})`
                }}
            >
            </div>

            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    {title}
                </p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{noteDate.format('dddd')}</span>
                <h5>{noteDate.format('D')} - {noteDate.format('MMM')}</h5>
            </div>
        </div>
    )
}
