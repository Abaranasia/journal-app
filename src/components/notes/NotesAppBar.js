import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active } = useSelector(state => state.notes);


    const handleSave = () => { // guarda una nota
        dispatch(startSaveNote(active))
    }

    const handlePictureClick = () => { //redirige el evento del envío de imagen
        document.querySelector('#fileSelector').click()
    }


    const handleFileChange = (e) => { //gestiona el envío de la imagen
        const file = (e.target.files[0]);
        console.log('FILE: ', file)
        if (file) { dispatch(startUploading(file)) }

    }
    return (
        <div className="notes__appbar">
            <span> 25 de septiembre de 2021</span>

            <input
                id="fileSelector"
                type="file"
                name="file"
                style={{ display: 'none' }}
                onChange={handleFileChange}
            />

            <div>
                <button className="btn" onClick={handlePictureClick}>
                    Picture
                </button>
                <button className="btn" onClick={handleSave}>
                    Save
                </button>
            </div>

        </div>
    )
}
