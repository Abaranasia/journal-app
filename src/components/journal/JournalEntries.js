import React from 'react'
import { useSelector } from 'react-redux';
import { JournayEntry } from './JournayEntry';

export const JournalEntries = () => {

    const { notes } = useSelector(state => state.notes); // obtenemos del store el listado de notes del usuario

    //console.log(notes);

    //const entries = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

    return (
        <div className="journal__entries ">

            {
                notes.map(note => (
                    <JournayEntry
                        key={note.id}
                        {...note}
                    />
                ))
            }

        </div>
    )
}
