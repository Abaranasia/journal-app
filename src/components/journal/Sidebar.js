import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { JournalEntries } from './JournalEntries';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';


export const Sidebar = () => {


    const dispatch = useDispatch();
    const { name } = useSelector(state => state.auth) // Esto nos permite extrar desde el state el nombre del usuario conectado

    const handleLogout = () => {
        dispatch(startLogout()) //invoca al action de logout en auth.js

    }

    const handleAddNew = () => {
        dispatch(startNewNote());

    }


    return (
        <aside className="journal__sidebar">

            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon"></i>
                    {name}
                </h3>
                <button className="btn" onClick={handleLogout}>Lougout</button>
            </div>

            <div className="journal__new-entry" onClick={handleAddNew}>
                <i className="far fa-calendar-plus fa-3x"></i>
                <p className="mt-5">New entry</p>
            </div>

            <JournalEntries />
        </aside>
    )
}
