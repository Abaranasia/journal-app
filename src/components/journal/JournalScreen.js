import React from 'react';
import { useSelector } from 'react-redux';
import { NoteScreen } from '../notes/NoteScreen';
import { NothingSelected } from './NothingSelected';
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {

    const { active: activeNote } = useSelector(state => state.notes); //para saber desde el reducer si la nota est√° o no activa

    return (
        <div className="journal__main-content animate__animated animate__fadeIn animate__faster">
            <Sidebar />

            <main>
                {
                    (activeNote)
                        ? (<NoteScreen />)
                        : (<NothingSelected />)
                }

            </main>
        </div>
    )
}
