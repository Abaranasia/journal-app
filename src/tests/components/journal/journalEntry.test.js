import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { JournayEntry } from '../../../components/journal/JournayEntry';
import { activeNote } from '../../../actions/notes';

//import { firebase } from '../../firebase/firebaseConfig';


//import { getAuth } from 'firebase/auth';
//const auth = getAuth();

//Evita el crasheo de tipo: INTERNAL ASSERTION FAILED: Expected a class definition en  firebase.auth()



const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Para evaluar estas pruebas vamos a necesitar que nos ayude el store. Para ello empleamos mockStore, 
// MockStore nos permitirá averiguarlas acciones que lo han disparado, con qué argumentos, etc.

const initState = {}

let store = mockStore(initState);
// Nuestro store fictício debe definir aquí el estado inicial en el que necesitamos que esté para estos test
store.dispatch = jest.fn(); // Mokeamos el dispatch porque necesitamos evaluarlo aquí

const note = {
    id: 123,
    date: 0,
    title: 'testing',
    body: 'body test',
    url: 'https://testing.com/foto.jpg'

}

const wrapper = mount(
    <Provider store={store}> {/* lo necesitamos para fingir el context del usuario */}
        <JournayEntry {...note} />
    </Provider>
)

describe('JournayEntry tests', () => {


    test('should render properly', () => {
        expect(wrapper).toMatchSnapshot()
    });


    test('should activate the note', () => {

        wrapper.find('.journal__entry').prop('onClick')();
        expect(store.dispatch).toHaveBeenCalled();
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(note.id, { ...note })
        );
    });
});