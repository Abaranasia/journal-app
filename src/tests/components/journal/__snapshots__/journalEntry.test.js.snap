// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournayEntry tests should render properly 1`] = `
Array [
  " ",
  <div
    className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
    onClick={[Function]}
  >
    <div
      className="journal__entry-picture"
      style={
        Object {
          "backgroundImage": "url(https://testing.com/foto.jpg)",
          "backgroundSize": "cover",
        }
      }
    />
    <div
      className="journal__entry-body"
    >
      <p
        className="journal__entry-title"
      >
        testing
      </p>
      <p
        className="journal__entry-content"
      >
        body test
      </p>
    </div>
    <div
      className="journal__entry-date-box"
    >
      <span>
        Thursday
      </span>
      <h5>
        1
         - 
        Jan
      </h5>
    </div>
  </div>,
]
`;
